src/unix.rs

```diff
@@ -33,20 +33,41 @@ fn _get(env: &impl EnvAccess) -> Option<String> {
         .filter(|val| !val.is_empty())
         .or_else(|| env.get(LANG))?;
 
-    parse_locale_code(&code)
+    Some(posix_to_bcp47(&code))
 }
 
-fn parse_locale_code(code: &str) -> Option<String> {
-    // Some locales are returned with the char encoding too, such as `en_US.UTF-8`
-    code.split_once('.')
-        .map(|(s, _)| s)
-        .or(Some(code))
-        .map(|s| s.replace('_', "-"))
+/// Converts a POSIX locale string to a BCP47 locale string.
+///
+/// This function processes the input `code` by removing any character encoding
+/// (the part after the first `.`) and any modifiers (the part after the first `@`).
+/// It replaces underscores (`_`) with hyphens (`-`) to conform to BCP47 formatting.
+///
+/// # Examples
+///
+/// ```do-not-compile
+/// let bcp47 = posix_to_bcp47("en_US.UTF-8");
+/// assert_eq!(bcp47, "en-US");
+///
+/// let bcp47 = posix_to_bcp47("fr_FR@dict");
+/// assert_eq!(bcp47, "fr-FR");
+///
+/// let bcp47 = posix_to_bcp47("de_DE.UTF-8@euro");
+/// assert_eq!(bcp47, "de-DE");
+/// ```
+///
+/// # TODO
+///
+/// Implement POSIX to BCP47 modifier conversion.
+fn posix_to_bcp47(code: &str) -> String {
+    code.chars()
+        .take_while(|&c| c != '.' && c != '@')
+        .collect::<String>()
+        .replace('_', "-")
 }
```

Comment:
Alternative implementation that can avoid heap allocation and make the replacement in_place:
```rust
fn posix_to_bcp47(code: &mut str) -> &str {
    let index = code.find(['@', '.']).unwrap_or(s.len());
    let sub = &mut code[..index];
    unsafe { sub.as_bytes_mut() }
        .iter_mut()
        .for_each(|c| {
            if *c == b'_' {
                *c = b'-';
            }
        });
    sub
}
```

Reply:
A shorter version for `String`:

```rust
fn posix_to_bcp47(mut code: String) -> String {
    code.find(['.', '@']).map(|i| code.truncate(i));
    unsafe { code.as_bytes_mut() }
        .iter_mut()
        .for_each(|c| if *c == b'_' { *c = b'-' });
    code
}
```

Reply:
This approach can also be implemented with the `String` type, making the function consuming.

```rust
fn posix_to_bcp47(mut code: String) -> String {
    let index = code.find(['.', '@']).unwrap_or(code.len());
    code.truncate(index);
    unsafe { code.as_bytes_mut() }
        .iter_mut()
        .for_each(|c| {
            if *c == b'_' {
                *c = b'-';
            }
        });
    code
}
```

Reply:
A shorter version for `String`:

```rust
fn posix_to_bcp47(mut code: String) -> String {
    code.find(['.', '@']).map(|i| code.truncate(i));
    unsafe { code.as_bytes_mut() }
        .iter_mut()
        .for_each(|c| if *c == b'_' { *c = b'-' });
    code
}
```

Reply:
And the same version for `&mut str`:

```rust
fn posix_to_bcp471(mut code: &mut str) -> &str {
    if let Some(i) = code.find(['.', '@']) {
        code = &mut code[..i];
    }
    unsafe { code.as_bytes_mut() }
        .iter_mut()
        .for_each(|c| if *c == b'_' { *c = b'-' });
    code
}
```

Reply:
In fact, I have concerns that this code may be incorrect in general, because some long characters can contain the byte `b'_'`, which will lead to unwanted byte replacement.
I suspect that such characters shouldn't be present in the locale code, but the concerns remain.

Also, editing the string in place will make using the function slightly more complicated in tests.
I think we can leave the current option (or, actually, my third option with `.map(|c| if c == '_' { '-' } else { c })`) for now and do optimization in the future.

Reply:
I've slightly changed the implementation.